### PR TITLE
Revert "[GSB] When adding same-type requirements pick representative based on canonical order"

### PR DESCRIPTION
**Pull request by DougGregor**
This reverts commit c725660cc9d69174809c214cf39d82fd8f807c33. It
uncovered a use-after-free in the GenericSignatureBuilder. Apologies
for the lack of a test case here; I'm still looking for something
small enough to commit.

Fixes rdar://problem/46772328.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: lib/AST/GenericSignatureBuilder.cpp [hunk 1]
```diff
@@ -4799,8 +4799,12 @@ GenericSignatureBuilder::addSameTypeRequirementBetweenTypeParameters(
...
   auto T2 = OrigT2->getRepresentative();
 
-  // Pick representative based on the canonical ordering of the type parameters.
-  if (compareDependentTypes(depType2, depType1) < 0) {
+  // Decide which potential archetype is to be considered the representative.
+  // We prefer potential archetypes with lower nesting depths, because it
+  // prevents us from unnecessarily building deeply nested potential archetypes.
+  unsigned nestingDepth1 = T1->getNestingDepth();
+  unsigned nestingDepth2 = T2->getNestingDepth();
+  if (nestingDepth2 < nestingDepth1) {
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply lib/AST/GenericSignatureBuilder.cpp]: Yes, that's the plan.

[the remark this answers is not in the document]
Comment (xedin) [reply lib/AST/GenericSignatureBuilder.cpp]: Thanks for clarification!

[the remark this answers is not in the document]
Comment (DougGregor) [reply lib/AST/GenericSignatureBuilder.cpp]: Yes.

**Pull request merged**